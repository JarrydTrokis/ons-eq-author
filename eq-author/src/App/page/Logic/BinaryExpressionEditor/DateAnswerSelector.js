import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { get, some } from "lodash";

import { colors, radius } from "constants/theme";
import { Number, Select, Label } from "components/Forms";
import VisuallyHidden from "components/VisuallyHidden";

import {
  rightSideErrors,
  OPERATOR_REQUIRED,
} from "constants/validationMessages";
import ValidationError from "components/ValidationError";

const conditions = {
  BEFORE: "Before",
  AFTER: "After",
  UNANSWERED: "Unanswered",
};

export const ConditionSelector = styled(Select)`
  width: 8em;
  flex: 1 1 auto;
  display: flex;
  position: relative;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  align-items: center;
  position: relative;
`;

const DateAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    outline-color: ${colors.errorPrimary};
    box-shadow: 0 0 0 2px ${colors.errorPrimary};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

const RuleText = styled.div`
  margin-left: 1em;
  margin-right: 1em;
`;

class DateAnswerSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      id: PropTypes.string.isRequired,
      left: PropTypes.shape({
        type: PropTypes.string.isRequired,
      }).isRequired,
      right: PropTypes.shape({
        offset: PropTypes.number,
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
    groupErrorMessage: PropTypes.string,
  };

  state = {
    offset: get(this.props.expression, "right.offset", null),
  };

  handleRightChange = ({ value }) => this.setState(() => ({ offset: value }));

  handleRightBlur = () => {
    this.props.onRightChange({
      dateValue: {
        offset: this.state.offset,
      },
    });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handleError = () => {
    const { expression, groupErrorMessage } = this.props;
    let message = null;

    const errors = expression.validationErrorInfo.errors;

    if (errors.some(({ field }) => field === "condition")) {
      message = OPERATOR_REQUIRED;
    } else if (
      errors.some(
        ({ errorCode }) =>
          errorCode === rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode
      )
    ) {
      message = rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.message;
    }

    if (
      some(expression.validationErrorInfo.errors, {
        errorCode: rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode,
      })
    ) {
      message = rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.message;
    }

    return <ValidationError>{message || groupErrorMessage}</ValidationError>;
  };

  render() {
    const { expression, groupErrorMessage } = this.props;

    const hasError =
      expression.validationErrorInfo.errors.length > 0 || groupErrorMessage;

    return (
      <>
        <DateAnswerRoutingSelector hasError={hasError}>
          <RuleText>is</RuleText>
          {expression.condition !== conditions.UNANSWERED &&
            expression.condition !== conditions.SELECT && (
              <>
                <Value>
                  <VisuallyHidden>
                    <Label htmlFor={`expression-right-${expression.id}`}>
                      Value
                    </Label>
                  </VisuallyHidden>
                  <Number
                    default={null}
                    id={`expression-right-${expression.id}`}
                    min={-99999999}
                    max={999999999}
                    placeholder="Value"
                    value={this.state.offset}
                    name={`expression-right-${expression.id}`}
                    onChange={this.handleRightChange}
                    onBlur={this.handleRightBlur}
                    data-test="number-value-input"
                    type="Number"
                    unit={get(expression.left, "properties.unit", null)}
                  />
                  <RuleText>years</RuleText>
                </Value>
              </>
            )}
          <VisuallyHidden>
            <Label htmlFor={`expression-condition-${expression.id}`}>
              Operator
            </Label>
          </VisuallyHidden>
          <ConditionSelector
            id={`expression-condition-${expression.id}`}
            onChange={this.handleConditionChange}
            name="condition-select"
            value={expression.condition}
            data-test="condition-selector"
          >
            <option value={conditions.BEFORE}> Before</option>
            <option value={conditions.AFTER}> After</option>
            <option value={conditions.UNANSWERED}> Unanswered</option>
          </ConditionSelector>
          <RuleText>response date</RuleText>
        </DateAnswerRoutingSelector>
        {hasError && this.handleError()}
      </>
    );
  }
}

export default DateAnswerSelector;
