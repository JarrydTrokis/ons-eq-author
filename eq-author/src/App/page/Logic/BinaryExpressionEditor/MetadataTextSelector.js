import React from "react";
import PropTypes from "prop-types";
import { useMutation } from "@apollo/react-hooks";
import styled from "styled-components";
import { colors, radius } from "constants/theme.js";

import WrappingInput from "components/Forms/WrappingInput";

import UPDATE_RIGHT_SIDE from "graphql/updateRightSide.graphql";

const RoutingSelectorContainer = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    outline-color: ${colors.errorPrimary};
    box-shadow: 0 0 0 2px ${colors.errorPrimary};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

const ConditionContent = styled.span`
  margin-right: 1em;
`;

const MetadataTextSelector = ({ expression }) => {
  const rightSideText = expression?.right?.text;
  const [updateRightSide] = useMutation(UPDATE_RIGHT_SIDE);

  return (
    <RoutingSelectorContainer>
      <ConditionContent>{expression.condition}</ConditionContent>
      <WrappingInput
        id="metadata-match-input"
        name="label"
        data-test="metadata-match-input"
        value={rightSideText}
        onChange={({ value }) =>
          updateRightSide({
            variables: {
              input: {
                expressionId: expression.id,
                customValue: { text: value },
              },
            },
          })
        }
      />
    </RoutingSelectorContainer>
  );
};

export default MetadataTextSelector;
