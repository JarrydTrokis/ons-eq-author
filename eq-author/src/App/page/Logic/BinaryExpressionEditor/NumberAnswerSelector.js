import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { get, some } from "lodash";

import { colors, radius } from "constants/theme";
import { Number, Select, Label } from "components/Forms";
import VisuallyHidden from "components/VisuallyHidden";

import { rightSideErrors } from "constants/validationMessages";
import { ValidationError } from "components/Error";

const conditions = {
  EQUAL: "Equal",
  NOT_EQUAL: "NotEqual",
  GREATER_THAN: "GreaterThan",
  LESS_THAN: "LessThan",
  GREATER_OR_EQUAL: "GreaterOrEqual",
  LESS_OR_EQUAL: "LessOrEqual",
  UNANSWERED: "Unanswered",
};

export const ConditionSelector = styled(Select)`
  width: auto;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  margin-left: 1em;
  align-items: center;
  position: relative;
`;

const NumberAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.red};
    outline-color: ${colors.red};
    box-shadow: 0 0 0 2px ${colors.red};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

class NumberAnswerSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      id: PropTypes.string.isRequired,
      left: PropTypes.shape({
        type: PropTypes.string.isRequired,
      }).isRequired,
      right: PropTypes.shape({
        number: PropTypes.number.isRequired,
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
  };

  state = {
    number: get(this.props.expression, "right.number", null),
  };

  handleRightChange = ({ value }) => this.setState(() => ({ number: value }));

  handleRightBlur = () => {
    this.props.onRightChange({ customValue: { number: this.state.number } });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handleError = () => {
    return (
      <ValidationError right>
        {rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.message}
      </ValidationError>
    );
  };

  render() {
    const { expression } = this.props;

    const hasError = some(expression.validationErrorInfo.errors, {
      errorCode: rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode,
    });

    return (
      <>
        <NumberAnswerRoutingSelector hasError={hasError}>
          <VisuallyHidden>
            <Label htmlFor={`expression-condition-${expression.id}`}>
              Operator
            </Label>
          </VisuallyHidden>
          <ConditionSelector
            id={`expression-condition-${expression.id}`}
            onChange={this.handleConditionChange}
            name="condition-select"
            value={expression.condition}
            data-test="condition-selector"
          >
            <option value={conditions.EQUAL}>(=) Equal to</option>
            <option value={conditions.NOT_EQUAL}>(&ne;) Not equal to</option>
            <option value={conditions.GREATER_THAN}>(&gt;) More than</option>
            <option value={conditions.LESS_THAN}>(&lt;) Less than</option>
            <option value={conditions.GREATER_OR_EQUAL}>
              (&ge;) More than or equal to
            </option>
            <option value={conditions.LESS_OR_EQUAL}>
              (&le;) Less than or equal to
            </option>
            <option value={conditions.UNANSWERED}>Unanswered</option>
          </ConditionSelector>
          {expression.condition !== conditions.UNANSWERED && (
            <>
              <Value>
                <VisuallyHidden>
                  <Label htmlFor={`expression-right-${expression.id}`}>
                    Value
                  </Label>
                </VisuallyHidden>
                <Number
                  default={null}
                  id={`expression-right-${expression.id}`}
                  min={-99999999}
                  max={999999999}
                  placeholder="Value"
                  value={this.state.number}
                  name={`expression-right-${expression.id}`}
                  onChange={this.handleRightChange}
                  onBlur={this.handleRightBlur}
                  data-test="number-value-input"
                  type={expression.left.type}
                  unit={get(expression.left, "properties.unit", null)}
                />
              </Value>
            </>
          )}
        </NumberAnswerRoutingSelector>
        {hasError && this.handleError()}
      </>
    );
  }
}

export default NumberAnswerSelector;
