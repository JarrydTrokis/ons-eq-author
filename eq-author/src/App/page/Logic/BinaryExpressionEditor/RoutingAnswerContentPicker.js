import React from "react";
import PropTypes from "prop-types";
import gql from "graphql-tag";
import { useQuery } from "@apollo/react-hooks";
import { get } from "lodash";
import { withRouter } from "react-router-dom";

import ContentPickerSelect from "components/ContentPickerSelect";
import { ANSWER } from "components/ContentPickerSelect/content-types";
import shapeTree from "components/ContentPicker/shapeTree";

const GET_AVAILABLE_ANSWERS = gql`
  query GetAvailableAnswers($input: GetAvailableAnswersInput!) {
    getAvailableAnswers(input: $input) {
      id
      displayName
      properties
      type
      page {
        id
        displayName
        section {
          id
          displayName
        }
      }
    }
  }
`;

export const RoutingAnswerContentPicker = ({
  match,
  includeSelf,
  path,
  ...otherProps
}) => {
  const { loading, error, data } = useQuery(GET_AVAILABLE_ANSWERS, {
    variables: {
      input: {
        pageId: match.params.confirmationId || match.params.pageId,
        includeSelf: Boolean(match.params.confirmationId) || includeSelf,
      },
    },
    fetchPolicy: "cache-and-network",
  });
  return (
    <ContentPickerSelect
      name="answerId"
      contentTypes={[ANSWER]}
      answerData={shapeTree(get(data, path))}
      loading={loading}
      error={error}
      {...otherProps}
    />
  );
};

RoutingAnswerContentPicker.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      confirmationId: PropTypes.string,
      pageId: PropTypes.string,
    }),
  }),
  includeSelf: PropTypes.bool,
  path: PropTypes.string.isRequired,
};

export default withRouter(RoutingAnswerContentPicker);
