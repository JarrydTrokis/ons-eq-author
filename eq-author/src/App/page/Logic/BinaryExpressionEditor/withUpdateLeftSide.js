import { graphql } from "react-apollo";

import updateLeftSide from "./updateLeftSide.graphql";

export const mapMutateToProps = ({ mutate }) => ({
  updateLeftSide(expression, answerId, metadataId) {
    return mutate({
      variables: {
        input: {
          expressionId: expression.id,
          answerId,
          metadataId,
        },
      },
      refetchQueries: ["GetQuestionnaire"],
    });
  },
});

export default graphql(updateLeftSide, {
  props: mapMutateToProps,
});
