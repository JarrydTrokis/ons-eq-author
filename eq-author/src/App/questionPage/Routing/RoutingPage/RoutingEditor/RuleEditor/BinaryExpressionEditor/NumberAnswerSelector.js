import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { get } from "lodash/fp";

import { colors, radius } from "constants/theme";
import { CURRENCY } from "constants/answer-types";
import { Number, Select, Label } from "components/Forms";
import VisuallyHidden from "components/VisuallyHidden";

export const RoutingNumberInput = styled(Number)`
  border-radius: ${radius};
  width: 8em;
  outline: none;
  ${props => props.answerType === CURRENCY && "padding-left: 1.2em;"}
`;

export const ConditionSelector = styled(Select)`
  width: auto;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  margin-left: 1em;
  align-items: center;
  position: relative;
`;

const CurrencySymbol = styled.div`
  position: absolute;
  opacity: 0.5;
  left: 0.5em;
`;

const NumberAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
`;

class NumberAnswerSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      id: PropTypes.string.isRequired,
      left: PropTypes.shape({
        type: PropTypes.string.isRequired,
      }).isRequired,
      right: PropTypes.shape({
        number: PropTypes.number.isRequired,
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
  };

  handleRightChange = ({ value }) => {
    this.props.onRightChange({ customValue: { number: value } });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  render() {
    const { expression } = this.props;
    return (
      <NumberAnswerRoutingSelector>
        <VisuallyHidden>
          <Label htmlFor={`expression-condition-${expression.id}`}>
            Operator
          </Label>
        </VisuallyHidden>
        <ConditionSelector
          id={`expression-condition-${expression.id}`}
          onChange={this.handleConditionChange}
          name="condition-select"
          value={expression.condition}
          data-test="condition-selector"
        >
          <option value="Equal">(=) Equal to</option>
          <option value="NotEqual">(&ne;) Not equal to</option>
          <option value="GreaterThan">(&gt;) More than</option>
          <option value="LessThan">(&lt;) Less than</option>
          <option value="GreaterOrEqual">(&ge;) More than or equal to</option>
          <option value="LessOrEqual">(&le;) Less than or equal to</option>
        </ConditionSelector>
        <Value>
          <VisuallyHidden>
            <Label htmlFor={`expression-right-${expression.id}`}>Value</Label>
          </VisuallyHidden>
          {expression.left.type === CURRENCY && (
            <CurrencySymbol>Â£</CurrencySymbol>
          )}
          <RoutingNumberInput
            id={`expression-right-${expression.id}`}
            min={-99999999}
            max={999999999}
            placeholder="Value"
            value={get("right.number", expression)}
            name={`expression-right-${expression.id}`}
            onChange={this.handleRightChange}
            data-test="number-value-input"
            answerType={expression.left.type}
          />
        </Value>
      </NumberAnswerRoutingSelector>
    );
  }
}

export default NumberAnswerSelector;
