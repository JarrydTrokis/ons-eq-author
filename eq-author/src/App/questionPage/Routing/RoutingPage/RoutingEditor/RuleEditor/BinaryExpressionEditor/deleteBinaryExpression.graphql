mutation deleteBinaryExpression2($input: DeleteBinaryExpression2Input!) {
  deleteBinaryExpression2(input: $input) {
    id
    expressionGroup {
      id
      expressions {
        ... on BinaryExpression2 {
          id
        }
      }
    }
  }
}
