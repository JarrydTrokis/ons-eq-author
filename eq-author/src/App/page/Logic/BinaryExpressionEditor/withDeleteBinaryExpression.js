import { graphql } from "react-apollo";

import deleteBinaryExpression from "./deleteBinaryExpression.graphql";

export const mapMutateToProps = ({ mutate }) => ({
  deleteBinaryExpression(id, onCompleted) {
    return mutate({
      variables: {
        input: {
          id,
        },
      },
      onCompleted
    });
  },
});

export default graphql(deleteBinaryExpression, {
  props: mapMutateToProps,
});
