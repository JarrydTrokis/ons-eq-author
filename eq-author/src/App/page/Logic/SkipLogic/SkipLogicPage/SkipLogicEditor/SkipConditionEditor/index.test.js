import React from "react";
import { shallow } from "enzyme";

import { RADIO } from "constants/answer-types";

import BinaryExpressionEditor from "App/page/Logic/BinaryExpressionEditor";

import { UnwrappedSkipConditionEditor as SkipConditionEditor } from "./";
import { byTestAttr } from "tests/utils/selectors";

describe("SkipConditionEditor", () => {
  let defaultProps;
  let defaultPropsMiddle;

  beforeEach(() => {
    defaultProps = {
      pageId: "1",
      expressionGroup: { id: "expGrpId", expressions: [] },
      expressionGroupIndex: 0,
      deleteSkipCondition: jest.fn(),
      deleteSkipConditions: jest.fn(),
    };
    defaultPropsMiddle = {
      pageId: "1",
      expressionGroup: { id: "expGrpId", expressions: [] },
      expressionGroupIndex: 1,
      deleteSkipCondition: jest.fn(),
      deleteSkipConditions: jest.fn(),
    };
  });

  it("should render", () => {
    const wrapper = shallow(<SkipConditionEditor {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should allow deleting all skip conditions", () => {
    const wrapper = shallow(<SkipConditionEditor {...defaultProps} />);
    wrapper.find(byTestAttr("btn-remove-skip-conditions")).simulate("click");
    expect(defaultProps.deleteSkipConditions).toHaveBeenCalledWith(
      defaultProps.pageId
    );
  });

  it("should allow deleting or skip conditions", () => {
    const wrapper = shallow(<SkipConditionEditor {...defaultPropsMiddle} />);
    wrapper.find(byTestAttr("btn-remove-skip-condition")).simulate("click");
    expect(defaultPropsMiddle.deleteSkipCondition).toHaveBeenCalledWith(
      defaultPropsMiddle.expressionGroup.id
    );
  });

  it("should pass down the correct prop when a second 'And' condition is invalid", () => {
    defaultProps.expressionGroup.expressions = [
      {
        id: "2",
        left: {
          id: "binaryExpressionId",
          type: RADIO,
        },
      },
      {
        id: "3",
        left: {
          id: "binaryExpressionId",
          type: RADIO,
        },
      },
    ];

    const wrapper = shallow(<SkipConditionEditor {...defaultProps} />);

    expect(
      wrapper
        .find(BinaryExpressionEditor)
        .first()
        .prop("canAddCondition")
    ).toBe(true);

    expect(
      wrapper
        .find(BinaryExpressionEditor)
        .last()
        .prop("canAddCondition")
    ).toBe(false);
  });
});
