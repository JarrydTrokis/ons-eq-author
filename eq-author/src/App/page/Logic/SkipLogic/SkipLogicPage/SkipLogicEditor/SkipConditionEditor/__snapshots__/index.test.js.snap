// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipConditionEditor should render 1`] = `
<SkipConditionEditor__SkipCondition
  data-test="skip-condition"
>
  <SkipConditionEditor__Header>
    <Label
      bold={true}
      inline={true}
    >
      Skip this quetion
    </Label>
    <SkipConditionEditor__RemoveRuleButton
      data-test="btn-remove-skip-conditions"
      onClick={[Function]}
      type="button"
      variant="primary"
    >
      Remove logic rule
    </SkipConditionEditor__RemoveRuleButton>
  </SkipConditionEditor__Header>
  <SkipConditionEditor__Expressions>
    <TransitionGroup
      childFactory={[Function]}
      component="div"
    />
  </SkipConditionEditor__Expressions>
</SkipConditionEditor__SkipCondition>
`;
